runtime/compiler/optimizer/J9RecognizedCallTransformer.cpp

```diff
@@ -77,6 +78,137 @@ void J9::RecognizedCallTransformer::process_java_lang_Class_IsAssignableFrom(TR:
    fromClass->recursivelyDecReferenceCount();
    }
 
+// This methods inlines a call node that calls StringCoding.encodeASCII into an if-diamond. The forward path of the
+// if-diamond inlines the calland the fallback path reverts back to calling the method in the traditional way.
+void J9::RecognizedCallTransformer::process_java_lang_StringCoding_encodeASCII(TR::TreeTop* treetop, TR::Node* node)
+   {
+   TR_J9VMBase *fej9 = static_cast<TR_J9VMBase*>(comp()->fe());
+   TR_OpaqueClassBlock *stringClass = fej9->getClassFromSignature("java/lang/String", 16, comp()->getCurrentMethod());
+   if (stringClass == NULL)
+      {
+      return;
+      }
+   uint32_t *latin1FieldAddress = (uint32_t *)fej9->getStaticFieldAddress(stringClass, (unsigned char *) "LATIN1", 6, (unsigned char *)"B", 1);
+   TR::CFG *cfg = comp()->getFlowGraph();
+
+   TR::Node *coderNode = node->getChild(0);
+   TR::Node *sourceArrayNode = node->getChild(1);
+
+   // Anchor the source array node as we need it in the fallthrough block.
+   anchorNode(sourceArrayNode, treetop);
+
+   // Now generate the ificmpne tree and insert it before the original call tree.
+   TR::Node *constNode = TR::Node::create(node, TR::iconst, 0, (uint32_t)*latin1FieldAddress);
+   TR::Node *ifCmpNode = TR::Node::createif(TR::ificmpne, coderNode, constNode);
+   TR::TreeTop *ifCmpTreeTop = TR::TreeTop::create(comp(), treetop->getPrevTreeTop(), ifCmpNode);
+
+   // Split the current block right before the call (or after the ificmpne).
+   TR::Block *ifCmpBlock = ifCmpTreeTop->getEnclosingBlock();
+   TR::Block *fallbackPathBlock = ifCmpBlock->split(treetop, cfg, true /* fixUpCommoning */, true /* copyExceptionSuccessors */);
+   // Then split again to create the tail block
+   TR::Block *tailBlock = fallbackPathBlock->split(treetop->getNextTreeTop(), cfg, true /* fixUpCommoning */, true /* copyExceptionSuccessors */);
+
+   // Now we have the originalBlock, the fallback block and the merge block (tail block). The original and fallback blocks have commoned nodes.
+   // The fallback block and tail block have commoned nodes as well. The call result will be an astore as well.
+   // If this worked as expected, then the treetop after the call node should be an astore to a temp  slot. So assert that it is, and then grab
+   // a reference to the temp slot so we can store to it ourselves in the forward path.
+   TR::Node *resultStoreNode = treetop->getNextTreeTop()->getNode();
+   TR::SymbolReference *tempSlotForCallResult = NULL;
+   if (resultStoreNode->getOpCode().getOpCodeValue() == TR::astore)
+      {
+      tempSlotForCallResult = resultStoreNode->getSymbolReference();
+      }
+   TR_ASSERT_FATAL(tempSlotForCallResult, "Symbol reference for store node can't be null\n");
+
+   // Ready to create our fallthrough block now. Connect it to ifCmpTreeTop, and then split it with nodes commoned.
+   int32_t byteArrayType = fej9->getNewArrayTypeFromClass(fej9->getByteArrayClass());
```

Comment:
Out of an abundance of caution (similar to checking if the String class exists) should we be checking if the byte array class exists before starting this optimization ?